### PR TITLE
Support for compositional Shrinking

### PR DESCRIPTION
**Pull request by mfelsche**
This is a **BIG** refactoring of how `Generator` and all its combinators work.

## Motivation

Previously the Api was roughly:

```pony

trait Generator[T]
  fun generate(rnd: Randomness): T^
  fun shrink(t: T): Seq[T^]
```

With this approach we called generate multiple times while testing a property. If the property failed for a sample `t` we called `shrink(t)` to obtain a sequence of shrunken samples that are in some way "smaller" than the sample `t`. Those sequences were quite short and every time a sample failed it was passed to `shrink` again to obtain even smaller examples.

This overall approach of having separate `generate` and `shrink` methods limits the shrinking substantially. If a generator is created using a chain of combinators (e.g. `Generators.u8().map({(u): U8 => u*2 })`) the shrinking is unable to properly incorporate the logic from the combinators as all that is available to shrink is the final result of the Combinators.

## API Change
With this pull request the Generator API changed to this:

```pony
type ValueAndShrink[T] is (T^, Iterator[T^])
type GenerateResult[T] is (T^ | ValueAndShrink[T])

trait Generator[T]
  fun generate(rnd: Randomness): GenerateResult[T]
  fun shrink(t: T): ValueAndShrink[T]
```

The basic message here is: Generators now **can** return a generated value and an Iterator for its shrinks from `generate`.

This will make the shrink values available for downstream Generator Combinators working on source Generators (like `Generator.map`, `Generators.zip2`, `Generator.filter` etc.), so they have their source iterators shrink values available for applying their logic on them to produce proper shrink values that are generated in the same way as the normally generated values.

This change introduces a whole lot of complexity for implementing Generators, making it harder for users to implement their own, it seems. But the type signature of `Generator` allows it to still write simple Generators in the old way (that do not support proper shrinking):

```pony
use "itertools"
use "ponycheck"

class MyGenerator is Generator[String]
  fun generate(rnd: Randomness): GenerateResult[String] =>
    rnd.u64().string()
  fun shrink(s: String): ValueAndShrink[String] =>
    let iter = 
      try
        Iter[U64](Range(s.u64()?, 0, -1).map[String]({(u: U64): String => u.string() })
      else
        // fallback
        Iter[String].repeat_value(s)
      end
      (consume s, iter)
```

### Examples

The new examples show how shrinking improved for a main use case of generating custom classes.

### CODE REVIEW

**Comment by mfelsche** [comment]
Travis seems to fail due to not being able to allocate during tests compilation.

**Comment by mfelsche** [comment]
Travis is running against the latest release of ponyc which is 0.20.0 and does not yet contain this fix: https://github.com/ponylang/ponyc/commit/c028049008e71733f971fa2fa0d100454943ec00 which we experience here due to this change: https://github.com/mfelsche/ponycheck/pull/27/files#diff-2aac35c3d6493b6049db3582d087c7edR1201

I added some workarounds, hoping this will fix it.

**Comment by Theodus** [comment]
I'm sorry you had to track that one down. It looks like the workaround works though.

**Comment by Theodus** [comment]
I'll probably review this in the morning

**Comment by mfelsche** [comment]
No problem at all. These bugs are already fixed so everything is fine. This took me so long, no need to hurry now. Take your time!

File: ponycheck/generator.pony [hunk 1]
```diff
@@ -1,68 +1,275 @@
...
+    apply `predicate` to the values generated by this Generator
+    and only values for which `predicate` returns `true`.
+
+    Example:
+
+    ```pony
+    let even_i32s =
+      Generators.i32()
+        .filter(
+          {(t: I32}: (I32^, Bool) => (t % 2) == 0 })
```
Comment (Theodus) [inline ponycheck/generator.pony]: The lambda in this example doesn't seem to return a tuple

File: ponycheck/generator.pony [hunk 2]
```diff
@@ -1,68 +1,275 @@
...
+    """
+    apply function `fn` to each value of this iterator
+    and yield the results.
+
+    Example:
+
+    ```pony
+    let single_code_point_string_gen =
+      Generators.u32()
+        .map[String]({(u: U32): String => String.from_utf32(u) })
```
Comment (Theodus) [inline ponycheck/generator.pony]: The examples with simple lambdas like this should probably use the new lambda syntax.
'''pony
.map[String]({(u) => String.from_utf32(u) })
'''

**Review by Theodus — CHANGES_REQUESTED** [review]

File: ponycheck/generator.pony (hunk 1 above)
Comment (mfelsche) [reply ponycheck/generator.pony]: my bad, corrected.

File: ponycheck/generator.pony (hunk 2 above)
Comment (mfelsche) [reply ponycheck/generator.pony]: Right, looks much nicer.
I adapted it everywhere possible.

**Review by Theodus — APPROVED** [review]

**Comment by Theodus** [comment]
Looks good to me!

**Pull request merged**